crypt-yescrypt.c

```diff
@@ -39,71 +50,105 @@ gensalt_yescrypt_rn(unsigned long count,
                     const uint8_t *rbytes, size_t nrbytes,
                     uint8_t *output, size_t o_size)
 {
-	/* Use one of recommended parameter sets as the 'low default'. */
-	yescrypt_params_t params = { .flags = YESCRYPT_DEFAULTS,
-		.N = 4096, .r = 32, .p = 1 };
-
-	if (count) {
-		/*
-		 * `1 << (count - 1)` is MiB usage in range of 1MiB..1GiB,
-		 * thus, count is in range of 1..11
-		 */
-		if (count <= 2) {
-			params.r = 8; /* N in 1KiB */
-			params.N = 512ULL << count;
-		} else if (count <= 11) {
-			params.r = 32; /* N in 4KiB */
-			params.N = 128ULL << count;
-		} else {
-			errno = EINVAL;
-			return;
-		}
-	}
-
-	if (!yescrypt_encode_params_r(&params, rbytes, nrbytes, output, o_size)) {
-		/*
-		 * As the output could have already been written,
-		 * overwrite it with a short failure token.
-		 */
-		output[0] = '*';
-		output[1] = '\0';
-		errno = ERANGE;
-		return;
-	}
+  if (o_size < 3 + 8 * 6 + BASE64_LEN (nrbytes) + 1 ||
+      CRYPT_GENSALT_OUTPUT_SIZE < 3 + 8 * 6 + BASE64_LEN (nrbytes) + 1)
+    {
+      errno = ERANGE;
+      return;
+    }
+
+  if (count > 11 || nrbytes < 16)
+    {
+      errno = EINVAL;
+      return;
+    }
+
+  /* Temporary buffer for operation.  The buffer is guaranteed to be
+     large enough to hold the maximum size of the generated salt and
+     is filled with '\0' bytes before use.  */
+  uint8_t outbuf[CRYPT_GENSALT_OUTPUT_SIZE];
+  XCRYPT_SECURE_MEMSET(outbuf, CRYPT_GENSALT_OUTPUT_SIZE);
+
+  yescrypt_params_t params =
+  {
+    .flags = YESCRYPT_DEFAULTS,
+    .p = 1,
+  };
+
+  /* Valid cost parameters are from 1 to 11.  The default is 5.
+     These are used to set yescrypt's 'N' and 'r' parameters as
+     follows:
+     N (block count) is specified in units of r (block size,
+     adjustable in steps of 128 bytes).
+
+     128 bytes * r = size of each memory block
+
+     128 bytes * r * N = total amount of memory used for hashing
+                         in N blocks of r * 128 bytes.
+
+     The author of yescrypt recommends in the documentation to use
+     r=8 (a block size of 1 KiB) for total sizes of 2 MiB and less,
+     and r=32 (a block size of 4KiB) above that.
+     This has to do with the typical per-core last-level cache sizes
+     of current CPUs.  */
+
+  if (count == 0)
+    count = 5;
+
+  if (count < 3)
+    {
+      params.r = 8;                   // N in 1KiB
+      params.N = 1ULL << (count + 9); // 1 -> 1024, 2 -> 2048
+    }
+  else
+    {
+      params.r = 32;                  // N in 4KiB
+      params.N = 1ULL << (count + 7); // 3 -> 1024, 4 -> 2048, ... 11 -> 262144
+    }
+
+  if (!yescrypt_encode_params_r(&params, rbytes, nrbytes, outbuf, o_size))
+    {
+      errno = ERANGE;
+      return;
+    }
+
+  memcpy(output, outbuf, o_size);
```

Comment:
FWIW, `memcpy (output, outbuf, strlen (outbuf) + 1);` is just a cryptic way to write `strcpy (output, outbuf);`, so if that's what we mean we should in fact use `strcpy()` and not try to hide it.

There's a trivial alternative to `strlcpy()` through using `strncat()` into a buffer with NUL in its first char, but we probably don't want a truncating function here, so if we do any copying like this at all (I'm out of context as to why we need(?) that), then `assert()` + `strcpy()` sounds appropriate.

Reply:
I wrote `assert` and then `memcpy` because it's likely to trip stupid automated linters that ding you for using `strcpy` at all.  Maybe we need another wrapper inline a la `XCRYPT_SECURE_MEMSET`: `XCRYPT_STRCPY_OR_ABORT(dest, destsize, src)`.

The reason for the _copy_ is that both `yescrypt_encode_params_r` and `yescrypt_r` might write to their output buffer even when they fail.  `outbuf` already has a failure token set up, so we want to preserve that until we know the process is successful.

Reply:
Maybe instead of copying, we should (re)introduce the failure token into `outbuf` once we know the process isn't successful?

Reply:
Another way to side-step these issues could be to be (re)introducing the failure token by a call to a function that always takes care of the `*0` / `*1` alternation.